File: macos/Markdown/Markdown/VaultStore.swift
//
//  VaultStore.swift
//  Markdown
//
//  Swift facade over the Rust `markdown_vault` library, which owns every read and write of
//  the notes folder along with its history. Mirrors VaultStore.cs on Windows.
//
//  Deliberately thin: one JSON call reaches every tool, so adding a tool is a change in
//  Rust alone. Path safety lives there too — nothing here validates a path, because
//  `confine.rs` already refuses anything outside the vault.
//

import Foundation

nonisolated enum VaultError: LocalizedError {
    /// The vault could not be opened at all.
    case unavailable(URL)
    /// The Rust side rejected the operation; the message is written for a person to read.
    case rejected(String)
    /// The reply was not the shape the ABI promises.
    case malformedReply

    var errorDescription: String? {
        switch self {
        case .unavailable(let url): "Could not open the vault at \(url.path)."
        case .rejected(let message): message
        case .malformedReply: "The vault returned an unexpected reply."
        }
    }
}

/// An open notes vault.
///
/// `nonisolated` because it is a handle wrapper with no shared mutable state of its own, and
/// because `deinit` has to be able to release the handle.
nonisolated final class VaultStore {
    private let handle: OpaquePointer

    /// Opens `root`, initialising history and recording a baseline commit on first use.
    init(root: URL) throws {
        guard let handle = root.path.withCString({ md_vault_open($0) }) else {
            throw VaultError.unavailable(root)
        }
        self.handle = handle
    }

    deinit {
        md_vault_close(handle)
    }

    // MARK: - The one call

    /// Runs a vault tool and returns its `result` object.
    @discardableResult
    func call(_ name: String, _ input: [String: Any] = [:]) throws -> [String: Any] {
        let arguments = try JSONSerialization.data(withJSONObject: input)
        guard let json = String(data: arguments, encoding: .utf8) else {
            throw VaultError.malformedReply
        }

        guard let raw = name.withCString({ name in
            json.withCString { md_vault_call(handle, name, $0) }
        }) else {
            throw VaultError.malformedReply
        }
        defer { md_string_free(raw) }

        let reply = String(cString: raw)
        guard let data = reply.data(using: .utf8),
              let object = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw VaultError.malformedReply
        }

        // A failing tool is an ordinary reply carrying a message worth showing.
        guard object["ok"] as? Bool == true else {
            throw VaultError.rejected(object["error"] as? String ?? "The vault refused that.")
        }
        return object["result"] as? [String: Any] ?? [:]
    }

    // MARK: - Typed conveniences

    func read(_ path: String) throws -> String {
        guard let content = try call("read_note", ["path": path])["content"] as? String else {
            throw VaultError.malformedReply
        }
        return content
    }

    /// Returns the commit id, or `nil` when the contents were already what was asked for.
    @discardableResult
    func write(_ path: String, contents: String) throws -> String? {
        try call("write_note", ["path": path, "content": contents])["commit"] as? String
    }

    @discardableResult
    func createFile(_ path: String, contents: String = "") throws -> String? {
        try call("create_note", ["path": path, "content": contents])["commit"] as? String
    }

    @discardableResult
    func createFolder(_ path: String) throws -> String? {
        try call("create_folder", ["path": path])["commit"] as? String
    }

    @discardableResult
    func move(from: String, to: String) throws -> String? {
        try call("move", ["from": from, "to": to])["commit"] as? String
    }

    @discardableResult
    func delete(_ path: String) throws -> String? {
        try call("delete", ["path": path])["commit"] as? String
    }

    @discardableResult
    func undo(commit: String) throws -> String? {
        try call("undo", ["commit": commit])["commit"] as? String
    }

    /// Relative path of `url` inside this vault, which is the only form the vault accepts.
    static func relativePath(of url: URL, in root: URL) -> String? {
        let target = url.standardizedFileURL.path
        let base = root.standardizedFileURL.path
        guard target.hasPrefix(base + "/") else { return nil }
        return String(target.dropFirst(base.count + 1))
    }
}

File: win/MarkdownWin/MarkdownWin/VaultStore.cs
//
//  VaultStore.cs
//  MarkdownWin
//
//  Managed facade over the Rust `markdown_vault` library, which owns every read and write of
//  the notes folder along with its history. Mirrors VaultStore.swift on macOS.
//
//  Deliberately thin: one JSON call reaches every tool, so adding a tool is a change in
//  Rust alone. Path safety lives there too — nothing here validates a path, because
//  confine.rs already refuses anything outside the vault.
//

using System;
using System.IO;
using System.Text.Json.Nodes;

namespace MarkdownWin;

internal enum VaultErrorKind
{
    /// The vault could not be opened at all.
    Unavailable,
    /// The Rust side rejected the operation; the message is written for a person to read.
    Rejected,
    /// The reply was not the shape the ABI promises.
    MalformedReply,
}

internal sealed class VaultException : Exception
{
    public VaultErrorKind Kind { get; }

    public VaultException(VaultErrorKind kind, string message) : base(message) => Kind = kind;
}

/// <summary>An open notes vault.</summary>
internal sealed class VaultStore : IDisposable
{
    private IntPtr handle;
    private bool disposed;

    /// <summary>Opens <paramref name="root"/>, initialising history and recording a baseline commit on first use.</summary>
    public VaultStore(string root)
    {
        handle = MarkdownCore.VaultOpen(root);
        if (handle == IntPtr.Zero)
        {
            throw new VaultException(VaultErrorKind.Unavailable, $"Could not open the vault at {root}.");
        }
    }

    public void Dispose()
    {
        if (disposed)
        {
            return;
        }

        disposed = true;
        MarkdownCore.VaultClose(handle);
        handle = IntPtr.Zero;
    }

    // MARK: - The one call

    /// <summary>Runs a vault tool and returns its `result` object.</summary>
    public JsonObject Call(string name, JsonObject? input = null)
    {
        ObjectDisposedException.ThrowIf(disposed, this);

        string inputJson = (input ?? new JsonObject()).ToJsonString();
        string? reply = MarkdownCore.VaultCall(handle, name, inputJson);
        if (reply is null)
        {
            throw new VaultException(VaultErrorKind.MalformedReply, "The vault returned an unexpected reply.");
        }

        JsonObject? envelope;
        try
        {
            envelope = JsonNode.Parse(reply) as JsonObject;
        }
        catch (System.Text.Json.JsonException)
        {
            envelope = null;
        }

        if (envelope is null)
        {
            throw new VaultException(VaultErrorKind.MalformedReply, "The vault returned an unexpected reply.");
        }

        bool ok = envelope["ok"]?.GetValue<bool>() ?? false;
        if (!ok)
        {
            string message = envelope["error"]?.GetValue<string>() ?? "The vault refused that.";
            throw new VaultException(VaultErrorKind.Rejected, message);
        }

        return envelope["result"] as JsonObject ?? new JsonObject();
    }

    // MARK: - Typed conveniences

    public string Read(string path)
    {
        JsonObject result = Call("read_note", new JsonObject { ["path"] = path });
        return result["content"]?.GetValue<string>()
            ?? throw new VaultException(VaultErrorKind.MalformedReply, "The vault returned an unexpected reply.");
    }

    /// <summary>Returns the commit id, or null when the contents were already what was asked for.</summary>
    public string? Write(string path, string contents) =>
        Call("write_note", new JsonObject { ["path"] = path, ["content"] = contents })["commit"]?.GetValue<string>();

    public string? CreateFile(string path, string contents = "") =>
        Call("create_note", new JsonObject { ["path"] = path, ["content"] = contents })["commit"]?.GetValue<string>();

    public string? CreateFolder(string path) =>
        Call("create_folder", new JsonObject { ["path"] = path })["commit"]?.GetValue<string>();

    public string? Move(string from, string to) =>
        Call("move", new JsonObject { ["from"] = from, ["to"] = to })["commit"]?.GetValue<string>();

    public string? Delete(string path) =>
        Call("delete", new JsonObject { ["path"] = path })["commit"]?.GetValue<string>();

    public string? Undo(string commit) =>
        Call("undo", new JsonObject { ["commit"] = commit })["commit"]?.GetValue<string>();

    /// <summary>Copies <paramref name="data"/> into the vault's `assets` folder, choosing a
    /// collision-free name derived from <paramref name="filename"/> rather than overwriting
    /// anything. Returns the vault-relative path actually used, and a best-effort MIME type
    /// for the caller to decide how to render it (e.g. inline image vs. a plain attachment
    /// link).</summary>
    public (string Path, string Mime) ImportAsset(string filename, byte[] data)
    {
        JsonObject result = Call("import_asset", new JsonObject
        {
            ["filename"] = filename,
            ["content_base64"] = Convert.ToBase64String(data),
        });
        string? path = result["path"]?.GetValue<string>();
        string? mime = result["mime"]?.GetValue<string>();
        if (path is null || mime is null)
        {
            throw new VaultException(VaultErrorKind.MalformedReply, "The vault returned an unexpected reply.");
        }

        return (path, mime);
    }

    /// <summary>Reads a vault file's raw bytes — unlike <see cref="Read"/>, not restricted to
    /// valid UTF-8, so it also serves images and other binary attachments.</summary>
    public (byte[] Data, string Mime) ReadAsset(string path)
    {
        JsonObject result = Call("read_asset", new JsonObject { ["path"] = path });
        string? contentBase64 = result["content_base64"]?.GetValue<string>();
        string? mime = result["mime"]?.GetValue<string>();
        if (contentBase64 is null || mime is null)
        {
            throw new VaultException(VaultErrorKind.MalformedReply, "The vault returned an unexpected reply.");
        }

        return (Convert.FromBase64String(contentBase64), mime);
    }

    /// <summary>Relative path of <paramref name="fullPath"/> inside <paramref name="root"/>, which is the
    /// only form the vault accepts. Returns null if <paramref name="fullPath"/> is not under <paramref name="root"/>.</summary>
    public static string? RelativePath(string fullPath, string root)
    {
        string target = Path.GetFullPath(fullPath);
        string baseDir = Path.GetFullPath(root).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        string prefix = baseDir + Path.DirectorySeparatorChar;

        return target.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)
            ? target[prefix.Length..]
            : null;
    }
}
